### ISSUE TITLE
Introduce a Generic Finite State Machine graph type

### ISSUE THREAD

**Issue by tarragonfly** [Vahera/godot-orchestrator#94]
### Description

- This would likely be a different graph type - State Graph. 
- States nest a script graph similar to a function/macro but has no input or output ports. EventGraph can be dragged into a State Graph as a new state.
- Introduce a transition node, which also nests a script graph, doesn't have input or output ports, but listens to signals emitted in connected State nodes.
- Transitions are done with signals by emitting a signal from a State node to a connected Transition node.
- Introduce two new events - On State Entered, On State Exit. 
- States can also contain any regular events such as Process.
- A new State Graph contains a single empty state that's marked as Start state, which is where the flow starts for this graph type.
- Add a new Node type called SubState(?) for EventGraph that nests a StateGraph with flow inputs for both starting and stopping the StateGraph

![+ Flowchart](https://github.com/Vahera/godot-orchestrator/assets/81910799/ce77b1df-28ba-43b3-8a41-ed9c04ea4e3e)
****

**Comment by Naros** [comment]
Hi @tarragonfly, I think, in general, this sounds like an elegant idea.  It would simplify our internal process, where we use another tool for Behavior Trees for enemy AI, and we certainly wanted to combine this technique somehow.

To provide some context on our internal process, we generally do this using Scene Node composition where we have a high-level scene node type that acts as our "AI" node that holds a blackboard, etc. Then as children of this node, we add and organize various control nodes like "Sequence" that then has a set of child behaviors.  There is a core FSM in the "AI" node, but it's more defined around the notion of behavior trees.

So I want to clarify a few points to make sure I understand your point of view.

A new graph type, `StateGraph` is meant to provide a slightly different canvas style for organizing game concepts, and in this case its a series of states and transitions. A state or transition can reference a single `EventGraph`, allowing the `StateGraph` to be a higher-level organization abstraction around a FSM; effectively reducing the overhead a user would need to implement to coordinate a FSM in visual scripts.

That said, what isn't immediately apparent to me from the description is how would the FSM ultimately be called from the outset.  Would you expect that an Orchestration Script (.os file) either be based on an `EventGraph` or a `StateGraph` or would the base "EventGraph" essentially call into one of your defined `StateGraph` instances that organizes various `EventGraph` instances into a FSM?

Since it seemed that your description indicates that an `EventGraph` can call into a `StateGraph` with a `SubState` node, that perhaps the outset is where an EventGraph calls into a specific `StateGraph` to initiate the FSM behavior, but perhaps I'm wrong?

**Comment by tarragonfly** [comment]
> A new graph type, `StateGraph` is meant to provide a slightly different canvas style for organizing game concepts, and in this case its a series of states and transitions. A state or transition can reference a single `EventGraph`, allowing the `StateGraph` to be a higher-level organization abstraction around a FSM; effectively reducing the overhead a user would need to implement to coordinate a FSM in visual scripts.
Correct.
> That said, what isn't immediately apparent to me from the description is how would the FSM ultimately be called from the outset. Would you expect that an Orchestration Script (.os file) either be based on an `EventGraph` or a `StateGraph` or would the base "EventGraph" essentially call into one of your defined `StateGraph` instances that organizes various `EventGraph` instances into a FSM?
Yea, it's not super clear in OP and I wasn't that clear on it when writing the feature request, but after contemplating on the issue I think State Graphs should fire immediately as soon as it's ready based on the marked Start state. The relevant event being On Enter State. There should be no manual initialization to streamline the system and lessen confusion. Also, rather than orchestration being either Event or State based, I think it would be better to have two graph types in the same.os. When you click the + button, you choose which to create from a dropdown. Graph types are differentiated by icon. EventGraph UI remains as is, but StateGraph UI links to dedicated States rather than events like EventGraph.
> Since it seemed that your description indicates that an EventGraph can call into a StateGraph with a SubState node, that perhaps the outset is where an EventGraph calls into a specific StateGraph to initiate the FSM behavior, but perhaps I'm wrong?

The nesting idea is starting to get confusing in general. Perhaps it's just better to have FSM be its own thing available as another option in parallel within the same .os but does not nest EventGraph type and vice versa.

**Comment by Naros** [comment]
> Correct, but I'm starting to waver on referencing existing EventGraphs as states. Perhaps it would be better to have dedicated States still expressed as EventGraphs but separate from current EventGraphs with its own dedicated StateGraph UI.

I completely agree. With this in mind, we've designed the UI to have a specific GraphEdit layout for when you want to create Macro Libraries and Function Libraries, so this would just be one additional layout.  I think this should be relatively easy to manage with the existing tab infrastructure.

> Yea, I wasn't that clear on it when writing the feature request, but after contemplating on the issue I think State Graphs should fire immediately as soon as they're ready from Start state. The relevant event being On Enter State. There should be no manual initialization to streamline the system and lessen confusion.

That sounds fair and definitely less confusing for the end user. 

> Nesting of different graph types is an interesting concept, but it's not that straightforward to implement or communicate about it. The idea stems from similar tools in Unity, but those are a lot more free-form and don't have a class/orchestration tying multiple concepts such as functions and signals together with several graphs. In those tools, there's just a single graph asset with a local variable blackboard and that's it. So nesting a graph asset inside another graph asset is not that complicated.

I also think that its more plausible for state and transitions to interact with functions and macros than they would with an event graph primarily on the basis that the interaction with Godot function overrides and callbacks are entirely abstracted away from the user in a StateGraph to guarantee the deterministic nature of the flow into and out of a state.

> There's also an interesting GDC AI talk from Naughty Dog, where they nest Behaviour Tree graphs inside FSM graphs for a hybrid approach in Uncharted games. But behaviour trees are out of scope for this feature request I think. FSM has a much wider application than just AI. And they are two different concepts in general.

Absolutely, though I think I would like to see that GDC talk to gain some insight nonetheless.  Thanks for the tidbit.

**Comment by tarragonfly** [comment]
> I also think that its more plausible for state and transitions to interact with functions and macros than they would with an event graph primarily on the basis that the interaction with Godot function overrides and callbacks are entirely abstracted away from the user in a StateGraph to guarantee the deterministic nature of the flow into and out of a state.
For sure, sounds good.
> Absolutely, though I think I would like to see that GDC talk to gain some insight nonetheless. Thanks for the tidbit.
Here's the talk: https://youtu.be/G8W7EQKBgcg?t=500

**Comment by eobet** [comment]
There are two different FSM officially in Unreal, yet also tonnes of branching "dialogue/quest" plug-ins in the market place which all do something very similar to FSM. And in my (Unreal) game I wanted again a similar functionality to drive my menus.

Wouldn't it be a USP for Orchestrator to have such a general FSM implementation that it can cover all of these cases? For example, I can't speak for dialog/quest implementations, but to have it work in a menu system the way I want, I'd need a transition node which has more than one exit point (ie one for each menu option in the current menu) so I can keep things tight and organized (having one transition node for each menu option in each menu will quickly become too large and unwieldly).
